File: 2022/Day_03/main.exs
defmodule Day_03 do
  @priorities (Enum.zip(Enum.concat(?a..?z, ?A..?Z), 1..52)
    |> Map.new)

  def parse_input(filename) do
    File.read!(filename)
    |> String.split("\n", trim: true)
    |> Enum.map(&String.to_charlist/1)
  end

  def part1(input) do
    input
    |> Enum.map(fn rucksack ->
      {compart1, compart2} = Enum.split(rucksack, div(length(rucksack), 2))
      MapSet.intersection(MapSet.new(compart1), MapSet.new(compart2))
      |> Enum.fetch!(0)
      |> then(& @priorities[&1])
    end)
    |> Enum.sum
  end

  def part2(input) do
    input
    |> Enum.chunk_every(3)
    |> Enum.map(fn s -> Enum.map(s, &MapSet.new/1) end)
    |> Enum.map(fn [sack1, sack2, sack3] ->
      MapSet.intersection(sack1, sack2)
      |> MapSet.intersection(sack3)
      |> Enum.fetch!(0)
      |> then(& @priorities[&1])
    end)
    |> Enum.sum
  end

  def main do
    filename = Enum.at(System.argv, 0, "input.txt")
    parsed_input = parse_input(filename)

    IO.inspect(part1(parsed_input), label: "Part 1")
    IO.inspect(part2(parsed_input), label: "Part 2")
  end
end

IO.inspect(:timer.tc(Day_03, :main, []) |> elem(0), label: "Time (uSecs)")

File: 2022/Day_03/main.hs
import qualified Data.Set as Set
import qualified Data.Map.Strict as Map
import Data.List.Split ( chunksOf )

import Control.Arrow ((>>>))
import Data.Function ((&))

priorities :: Map.Map Char Integer
priorities = Map.fromList (zip (['a'..'z'] ++ ['A'..'Z']) [1..52])

parseInput :: String -> IO [String]
parseInput filename = lines <$> readFile filename


part1 :: [String] -> Integer
part1 = 
    map (\sack -> 
        let half = length sack `div` 2
            (compart1, compart2) = splitAt half sack in 
        Set.intersection (Set.fromList compart1) (Set.fromList compart2)
        & Set.elemAt 0
        & (Map.!) priorities)
    >>> sum


part2 :: [String] -> Integer
part2 = 
    chunksOf 3
    >>> map (map Set.fromList)
    >>> map (\[sack1, sack2, sack3] ->
        Set.intersection sack1 sack2
        & Set.intersection sack3
        & Set.elemAt 0
        & (Map.!) priorities)
    >>> sum

main = do
    sacks <- parseInput "input.txt"
    putStrLn $ "Part 1: " ++ show (part1 sacks)
    putStrLn $ "Part 2: " ++ show (part2 sacks)

